File: NowPlaying/Android/src/org/metasyntactic/data/Performance.java
//Copyright 2008 Cyrus Najmabadi
//
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.
package org.metasyntactic.data;

import android.os.Parcel;
import android.os.Parcelable;
import org.metasyntactic.io.AbstractPersistable;
import org.metasyntactic.io.PersistableInputStream;
import org.metasyntactic.io.PersistableOutputStream;

import java.io.IOException;

public class Performance extends AbstractPersistable implements Parcelable {
  private static final long serialVersionUID = -3891926085016033570L;
  /*
  private final static SimpleDateFormat format = new SimpleDateFormat("h:mma");

  static {
    final DateFormatSymbols symbols = format.getDateFormatSymbols();
    symbols.setAmPmStrings(new String[]{"am", "pm"});
    format.setDateFormatSymbols(symbols);
  }
  */

  private final String time;
  private final String url;

  public void persistTo(final PersistableOutputStream out) throws IOException {
    out.writeString(time);
    out.writeString(url);
  }

  public static final Reader<Performance> reader = new AbstractReader<Performance>() {
    public Performance read(final PersistableInputStream in) throws IOException {
      final String time = in.readString();
      final String url = in.readString();

      return new Performance(time, url);
    }
  };

  public Performance(final String time, final String url) {
    this.time = time;
    this.url = url;
  }

  public String getTime() {
    return time;
  }

  public String getUrl() {
    return url;
  }

  public int describeContents() {
    return 0;
  }

  public void writeToParcel(final Parcel dest, final int flags) {
    dest.writeValue(time);
    dest.writeString(url);
  }

  public static final Creator<Performance> CREATOR = new Creator<Performance>() {
    public Performance createFromParcel(final Parcel source) {
      final String time = source.readString();
      final String url = source.readString();
      return new Performance(time, url);
    }

    public Performance[] newArray(final int size) {
      return new Performance[size];
    }
  };
}

File: tags/NowPlaying/iPhone/3.1/Classes/Performance.m
// Copyright 2008 Cyrus Najmabadi
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "Performance.h"

#import "DateUtilities.h"
#import "Utilities.h"

@interface Performance()
@property (copy) NSDate* time;
@property (copy) NSString* url;
@property (copy) NSString* timeString;
@end

@implementation Performance

property_definition(time);
property_definition(url);
@synthesize timeString;

- (void) dealloc {
    self.time = nil;
    self.url = nil;
    self.timeString = nil;

    [super dealloc];
}


- (id) initWithTime:(NSDate*) time_
                url:(NSString*) url_ {
    if (self = [super init]) {
        self.time = time_;
        self.url = [Utilities nonNilString:url_];
    }

    return self;
}


- (id) initWithCoder:(NSCoder*) coder {
    return [self initWithTime:[coder decodeObjectForKey:time_key]
                          url:[coder decodeObjectForKey:url_key]];
}


+ (Performance*) performanceWithTime:(NSDate*) time
                                 url:(NSString*) url {
    return [[[Performance alloc] initWithTime:time
                                          url:url] autorelease];
}


+ (Performance*) performanceWithDictionary:(NSDictionary*) dictionary {
    return [Performance performanceWithTime:[dictionary valueForKey:time_key]
                                        url:[dictionary valueForKey:url_key]];
}


- (NSDictionary*) dictionary {
    NSMutableDictionary* dictionary = [NSMutableDictionary dictionary];

    [dictionary setObject:time forKey:time_key];
    [dictionary setObject:url forKey:url_key];

    return dictionary;
}


- (void) encodeWithCoder:(NSCoder*) coder {
    [coder encodeObject:time forKey:time_key];
    [coder encodeObject:url forKey:url_key];
}


- (id) copyWithZone:(NSZone*) zone {
    return [self retain];
}


- (NSString*) timeString {
    if (timeString == nil) {
        self.timeString = [DateUtilities formatShortTime:time];
    }

    return timeString;
}

@end